docs: Add tutorial on encrypted events and examples for Sapphire (#612)
oasisprotocol/sapphire-paratime: docs/develop/encrypted-events.mdx

    // Implement your own guard (e.g., allowlist, OPL/ROFL attestation checks, etc.)
    modifier onlyROFL() {
        // TODO: enforce policy appropriate for your deployment
        _;
    }

    function emitWithOnchainKey(bytes calldata text) external onlyROFL {
        // 1) Generate a fresh symmetric key on-chain (per call or per session)
        bytes32 key = bytes32(Sapphire.randomBytes(32, bytes("rofl:onchain:key")));

        // 2) Encrypt with optional context-binding AAD
        bytes32 nonce = bytes32(Sapphire.randomBytes(32, bytes("rofl:nonce")));
        bytes memory ad = abi.encodePacked(block.chainid, address(this));
        bytes memory encrypted = Sapphire.encrypt(key, nonce, text, ad);

        // 3) Emit
        emit Encrypted(msg.sender, nonce, encrypted);

        // Optionally: persist or wrap `key` for ROFL retrieval using your chosen scheme.
        // Do NOT emit or log secrets in production.
    }
}
```

Notes

- Keep the key **inside** the contract/state or wrap it for ROFL; never emit it.
- Use AAD to bind ciphertexts to contract/chain or sender.
- If deriving ROFL-side keys, ensure the derivation mirrors your on-chain logic.

## Full Example and Demo

> :::example Encrypted Events

Reviewer: ```suggestion
:::example Encrypted Events

See the **[Encrypted Events Demo][encrypted-events-example]** repository for end-to-end Hardhat tasks to deploy, emit, and decrypt events using the patterns above, plus tests and a step-by-step tutorial.

:::
```

   public function intended for ROFL-based consumers.

We'll cover the on-chain smart contract logic and the off-chain decryption
process. A full working example is available inside the
**[Sapphire Examples][encrypted-events-example]** folder.

## How it Works: Basic Concept

To emit the encrypted event you need to **define an event**,
**generate a nonce**, **encrypt the text** with
additional authenticated data (AAD), and finally **emit the ciphertext**.
How you obtain the key is covered in the specific pattern subsection below.

### The `Encrypted` Event

Define a standard event signature in your contract. This structure is shared by
all patterns.

```solidity
// In your contract
event Encrypted(address indexed sender, bytes32 nonce, bytes ciphertext);
```

- `sender`: The address that initiated the encryption. Indexing it helps in
  filtering events.
- `nonce`: A unique number used once per encryption with a given key. **Never
  reuse a key/nonce pair!**
- `ciphertext`: The encrypted data payload.

### On-Chain Encryption

Sapphire provides a precompile for encryption. The basic flow inside a contract
function is:

> 1. **Get a Nonce:** Generate a fresh, random nonce for each encryption. Using a

Reviewer: ```suggestion
1. **Get a Nonce:** Generate a fresh, random nonce for each encryption with [`Sapphire.randomBytes`]. Passing a
```

And add a link to the API as below.

   domain separator is good practice.

   ```solidity
   // A 32-byte nonce for Deoxys-II, which uses the first 15 bytes
   bytes32 nonce = bytes32(Sapphire.randomBytes(32, bytes("my-dapp-nonce")));
   ```

> 2. **Encrypt the Data:** Use `Sapphire.encrypt` with a key, the nonce, your

Reviewer: ```suggestion
2. **Encrypt the Data:** Use [`Sapphire.encrypt`] with a key, the nonce, your
```

Then below, add
```markdown
[`Sapphire.encrypt`]: https://api.docs.oasis.io/sol/sapphire-contracts/contracts/Sapphire.sol/library.Sapphire.html#encrypt-1
```
